docs/manual/src/interface/objects.md

```diff
@@ -0,0 +1,281 @@
+
+```rust
+#[uniffi::declare_interface]
+mod example {
+    struct TodoList {
+        items: Vec<String>
+    }
+
+    impl TodoList {
+        fn new() -> Self {
+            TodoList {
+                items: Vec::new()
+            }
+        }
+
+        fn add_item(&mut self, todo: String) {
+            self.items.push(todo);
+        }
+
+        fn get_items(&self) -> Vec<String> {
+            self.items.clone()
+        }
+    }
+}
+```
+
+Conceptually, these objects are live Rust objects that have a proxy on the foreign language side; calling any methods on them, including a constructor or destructor results in the corresponding methods being called in Rust.
+
+UniFFI will generate these proxies of live objects with an interface or protocol. For example,
+in Kotlin the `TodoList` object would produce:
+
+
+```kotlin
+interface TodoListInterface {
+    fun addItem(todo: String)
+    fun getItems(): List<String>
+}
+
+class TodoList : TodoListInterface {
+   // implementations to call the Rust code.
+}
+```
+
+When working with these objects, it may be helpful to always pass the interface or protocol, but construct the concrete implementation. For example, in Swift:
+
+```swift
+let todoList = TodoList()
+todoList.addItem(todo: "Write documentation")
+display(list: todoList)
+
+func display(list: TodoListProtocol) {
+    let items = list.getItems()
+    items.forEach {
+        print($0)
+    }
+}
+```
+
+## Alternate Named Constructors
+
+In addition to the default constructor connected to the `::new()` method, you can specify
+alternate named constructors to create object instances in different ways. These are
+declared in the Rust code as methods without a `self` argument and returning an owned
+instance of `Self`:
+
+```rust
+    impl TodoList {
+        // The default constructor makes an empty list.
+        fn new() -> Self {
+            TodoList {
+                items: Vec::new()
+            }
+        }
+
+        // This alternate constructor makes a new TodoList from a list of string items.
+        fn new_from_items(items: Vec<String>) -> Self {
+            let mut obj = Self::new();
+            for item in items {
+                obj.add_item(item);
+            }
+            obj
+        }
+    }
+```
+
+For each alternate constructor, UniFFI will expose an appropriate static-method, class-method or similar
+in the foreign language binding, and will connect it to the Rust method of the same name on the underlying Rust struct.
+
+
+## Managing Shared References
+
+To the foreign-language consumer, UniFFI object instances are designed to behave as much like
+regular language objects as possible. They can be freely passed as arguments or returned as values,
+like this Kotlin API:
+
+```kotlin
+interface TodoListInterface {
+    // Copy the items from another TodoList into this one.
+    // TODO: hrm, actually we need to pass a concrete `TodoList` instance here,
```

Comment:
I'm a bit confused about this comment. I assumed that in the swift example above, it is recommended that the protocol (`TodoListProtocol`) be passed to a function instead of the concretion (`TodoList`) because it allows for increased flexibility so I'm not sure why it is a bad idea here.

Reply:
What I really want here is a `Self` type parameter. If you have some Swift-level mock implementation of the `TodoListInterface` then you absolutely cannot pass that in to the Rust code and expect it to work out well. The Rust code will only accept instances of the concrete `TodoList` type.

Reply:
I'm inclined to think that this "shared references" section could move to a more "advanced" section - this also confused me, and the `Arc<>` discussion still made me stop and think, so someone new to UniFFI might be best served by not encountering this here.

(OTOH, the following "concurrent access" section did feel OK in this doc, mainly because it's something you really can't ignore - whereas this section *can* be ignored until you actually need it?